### ISSUE TITLE
Add  -UnifyProperties parameter to Select-Object

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#13906]
Not sure if `-UnifyProperties` is the correct name for the parameter purpose described below.
Anyways, it would be nice to have an easy and _standard_ way to resolve the (common) issue were properties aren't displayed or taken by the next cmdlet because the first object in the pipeline doesn't contain all the properties of the following objects.

See e.g.: [Not all properties displayed](https://stackoverflow.com/q/44428189/1701026) and I believe just a new StackOverflow question came in with the same cause: [Trying to get all Teams with their owners, members and guest in a CSV using Powershell](https://stackoverflow.com/q/64550092/1701026).

## Proposed technical implementation details
`Select-Object -UnifyProperties` should behave similar to the purposed `Union-Object` function (v0.2.1) described in the [Not all properties displayed answer](https://stackoverflow.com/a/44429084/1701026). Meaning that it is expected to stall the pipeline to collect *all* properties names of all objects in the pipeline and use them as a property selection

### Related
With the purpose in place, the properties of the issues below could be obtained by simply piping the object to `| Select-Object -UnifyProperties`:
* **format-table should at least warn when it doesn't display properties** #7871
```PowerShell
[pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'} | Select-Object -UnifyProperties
```

```
name address phone
---- ------- -----
joe  home
             1
```

* **Cannot display two or more variables successively when "Select-Object" used** #12825
```PowerSehll
$Adapters, $IPAddress | Select-Object -UnifyProperties
```
```PowerSehll
$printers, $process | Select-Object -UnifyProperties
```

**Comment by vexx32** [comment]
`-Union` seems to imply objects are joined, when in reality you're not really adding data to any objects, just padding out extra properties so the display format is clear. 

I do agree it would be helpful for Select-Object to have this kind of functionality, but I don't think the parameter name communicates the intent very well. 🤔

**Comment by iRon7** [comment]
Agree, I am not native English, and I have no clue what should be a correct name, any suggestions?

**Comment by vexx32** [comment]
I'm not sure myself really. `-StandardizeProperties` would _maybe_ be ok but it feels like there should be a simpler term for it that I can't find at the moment. 😅

**Comment by iRon7** [comment]
I have changed the title/content using ` -UniteProperties` for now as `-Union` is definitely wrong and confusing.

**Comment by scriptingstudio** [comment]
It is all about `join-object`, is it not it?

**Comment by vexx32** [comment]
Typically I see Join-Object mentioned where you have two discrete sets of objects that you want to combine together based on a shared property/properties. My understanding of this suggestion (correct me if I'm wrong, please, @iRon7 🙂) is that we're mainly concerned with ensuring all the objects have the same property names, rather than merging related objects?

**Comment by iRon7** [comment]
Correct, it is **_not_** something like [`Join-Object`](https://github.com/iRon7/Join-Object), it is what @vexx32 describes: **ensuring all the objects have the same property names**

**Comment by iRon7** [comment]
It might also be considered to give this feature to a specific wildcard property (e.g. a double asterisk: `-Property **`):
```PowerShell
[pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'} | Select-Object **

name address phone
---- ------- -----
joe  home
             1
```

**Comment by scriptingstudio** [comment]
@iRon7 your example is exactly what any implementation of `join-object` does including my own. So I vote for a new native cmdlet `join-object`.

**Comment by vexx32** [comment]
Does it? Most Join-Object implementations I've seen would result in a single object. @iRon7's example is still two objects.

**Comment by scriptingstudio** [comment]
Single object is a result of merge. Join just adds object to another one. That is my understanding.

**Comment by vexx32** [comment]
That's not what's happening here, either? Neither object is being added to the other. Just ensuring that all the objects have the same set of property names.

**Comment by mklement0** [comment]
I like the idea in general, but we need to distinguish between whether this is for _display formatting_ or really about _creating objects_ that have the union of all properties across all input objects.

If this is about display formatting, then `Select-Object` isn't the right cmdlet to extend - `Format-Table` would be, perhaps with an `-AllProperties` switch:

Here's a quick prototype: 

```powershell
function Format-TableAllProperties {
  [System.Collections.Generic.List[string]] $propNames = @()
  [System.Collections.Generic.HashSet[string]] $hashSet = @()
  $inputCollected = @($input)
  $inputCollected.ForEach({ 
    foreach ($name in $_.psobject.Properties.Name) {
      if ($hashSet.Add($name)) { $propNames.Add($name) }
    }
  })
  $inputCollected | Format-Table $propNames
}
```

```powershell
PS> [pscustomobject] @{ one = 1; two = 2; three = 3 }, [pscustomobject] @{ one = 10; three = 30; four = 4 } | 
      Format-AllProperties

one two three four
--- --- ----- ----
  1   2     3 
 10        30 4
```

**Comment by jhoneill** [comment]
@dfinke had to put `Update-FirstObjectProperties` [into ImportExcel](https://github.com/dfinke/ImportExcel/blob/master/Public/Update-FirstObjectProperties.ps1)  for exactly this reason.  There must be a stack of places which need it.  

Whether adding it to `Select-object` is better than having a self contained command can be argued both ways. 
What that function could be implemented in a proxy command wrapping `Select-Object` but I've grown to expect it to be in its own command, so that's my bias, but I'm also thinking `Select-Object `is one of those widely used cmdlets that people might prefer to leave alone.

**Comment by mklement0** [comment]
So it sounds like _both_ a for-display and an extend-actual-objects solution may be desirable (I suspect you're aware of it, @jhoneill, but just to make it explicit: the linked function uses only the _first_ input object as the source for the set of properties whose presence should be ensured on all subsequent ones).

@iRon7, can we get clarity on which one you were looking for - the examples in the OP suggest the former - and perhaps create a _separate_ issue for the respective other, or at least clearly distinguish  these use cases.

**Comment by iRon7** [comment]
@mklement0,

> for _display formatting_ or really about _creating objects_

It is about *creating objects* (at least from my view), I had more something in mind like:

```PowerShell
function UniteProperties {                               # Select-Object -UniteProperties
  [System.Collections.Generic.List[string]] $propNames = @()
  [System.Collections.Generic.HashSet[string]] $hashSet = @()
  $inputCollected = @($input)
  $inputCollected.ForEach({ 
    foreach ($name in $_.psobject.Properties.Name) {
      if ($hashSet.Add($name)) { $propNames.Add($name) }
    }
  })
  $inputCollected | Select-Object $propNames
}
```

Current situation:
```PowerShell
[pscustomobject] @{ one = 1; two = 2; three = 3 },
[pscustomobject] @{ one = 10; three = 30; four = 4 } |
    ConvertTo-Csv

"one","two","three"
"1","2","3"
"10",,"30"
```

Future situation
```
[pscustomobject] @{ one = 1; two = 2; three = 3 },
[pscustomobject] @{ one = 10; three = 30; four = 4 } |
    UniteProperties | ConvertTo-Csv              # Select-Object -UniteProperties | ConvertTo-Csv

"one","two","three","four"
"1","2","3",
"10",,"30","4"
```

**Comment by iRon7** [comment]
> (I suspect you're aware of it , @jhoneill, and the function name suggests it, but just to make it explicit: the linked function uses only the _first_ input object as the source for the set of properties whose presence should be ensured on all subsequent ones).

Yes, it should be on all subsequent objects, because if you correct only the first object and then `... | Sort-Object | ...`, it might lose some properties again (see the description in the [Not all properties displayed answer](https://stackoverflow.com/q/64550092/1701026).

**Comment by mklement0** [comment]
Thanks for clarifying, @iRon7.

As for the parameter name, maybe `-UnifyProperties` is better?

As for a potential separate cmdlet instead: I struggle to even think of a good name, because there is no fitting approved verb that I see (there's some conceptual similarity to `Add-Member`, but adding it to that is worse, I think).

I think adding this to `Select-Object` is a good fit in terms of user expectations, even though the collect-all-input-up-front behavior is a departure, but addressing that through the help should suffice.

However, I don't know if there are implementation challenges and we would need to decide whether want to allow combining the new switch with any of the existing sub-selection functionality (`-First`, `-Skip`, `-Unique` ...)

**Comment by jhoneill** [comment]
> (I suspect you're aware of it , @jhoneill, and the function name suggests it, but just to make it explicit: the linked function uses only the _first_ input object as the source for the set of properties whose presence should be ensured on all subsequent ones).

There are two situations, one is where the first object determines what will be displayed / exported and the other objects can safely be left with a subset of the fields.  
Someone wedded to strict mode might hit problems with properties missing, so I wouldn't rule out ensuring presence, but
* adding many properties is slow compared to checking for their presence.
*  A piece of code which assumes properties are present is also likely to assume that one is a string, one is an integer, on is a Boolean etc, so the properties must be typed correctly
* If the type is bool or in and not  the nullable version one can't expect the receiving software to accept null.
* Giving fields a value (zero, false, empty string) changes the data from "Not known" to "known and zero/false/empty" which can have unpredictable consequences. 
Like I say I wouldn't argue there's no need, but this a dangerous thing to do. When you want to ensure a _known_ set of properties are all present on all objects, you just run through Select -property a,b,.y,z . 
The use case is (surely?) that the properties can't be known in advance, but you want to see all of them where they exist. Not that you want to fill in missing columns in each row with fictional values.

**Comment by jhoneill** [comment]
> It is about _creating objects_ (at least from my view), I had more something in mind like:
> 
> ```powershell
> function UniteProperties {                               # Select-Object -UniteProperties
>   [System.Collections.Generic.List[string]] $propNames = @()
>   [System.Collections.Generic.HashSet[string]] $hashSet = @()
>   $inputCollected = @($input)
>   $inputCollected.ForEach({ 
>     foreach ($name in $_.psobject.Properties.Name) {
>       if ($hashSet.Add($name)) { $propNames.Add($name) }
>     }
>   })
>   $inputCollected | Select-Object $propNames
> }
> ```

Or more simply, and more powershell-styled. 
```powershell
function UniteProperties {
    $hash =@{}
    $i = @($input)
    foreach ($obj in $i) {foreach ($p in $obj.psobject.properties) {$hash[$p.name] = $true} }
    $i | Select-Object ($hash.keys | ForEach-Object tostring)
}


$y [pscustomobject] @{ one = 1; two = 2; three = 3 },
[pscustomobject] @{ one = 10; three = 30; four = 4 } |
    UniteProperties 

$y | convertto-csv

$y[1].four.gettype()

gm -in $y[0] 

$y[0].four.gettype()
```

Have a look at what the last three lines do. That's part of what I was trying to explain to @mklement0  and probably didn't make sense.  That maybe what you want ...

**Comment by iRon7** [comment]
@jhoneill,
I guess you meant: `$y = [pscustomobject] ...`
In that case, all the examples behave as I would expect:

**`$y | convertto-csv `**
All object are converted to a csv format. Where every property (independent of its type) is wrapped double quotes, except for `$Null` with is left empty:
```PowerShell
"one","two","three","four"
"1","2","3",
"10",,"30","4"
```

**`$y[1].four.gettype()`**
The property of the second object (`[1]`) is set to an integer (`four = 4`):
```
IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Int32                                    System.ValueType
```

**`gm -in $y[0]`**
I am not familiar to this syntax, but assume it is similar to `$y[0] | gm`.
The properties `One`, `Two`, and `Three` are set to an integer (as above) and the property `four` is `null` (where `Get-Member` shows a general `Object` for the MemberType which unrelated to this purpose)
```
Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
four        NoteProperty object four=null
one         NoteProperty int one=1
three       NoteProperty int three=3
two         NoteProperty int two=2
```

**`$y[0].four.gettype()`**
Is resulting in an error as the value is not supplied (`$Null`), similar to: `$Null.GetType()`
```
InvalidOperation: You cannot call a method on a null-valued expression.
```
The results are that same if you do:
```PowerShell
$y = [pscustomobject] @{ one = 1; two = 2; three = 3 },
[pscustomobject] @{ one = 10; three = 30; four = 4 } |
    UniteProperties | Select-Object one, two, three, four
```
And similar to just:
```PowerShell
$y = [pscustomobject] @{ one = 1; two = 2; three = 3 },
[pscustomobject] @{ one = 10; three = 30; four = 4 }
```

**Comment by iRon7** [comment]
@mklement0,

> As for the parameter name, maybe -UnifyProperties is better?

I am fine with this, although I did have some more thoughts about this. Rather than choosing a _verb_ for a parameters, rely on the the cmdlet's verb, which results in something like this: `... | Select -EveryProperty` or simply  `... | Select -AllProperties`

Which also shows why I think it should be a feature of `Select-Object`. Besides, `Select-Object` has already a similar output if you define the properties yourself, like `...  | Select-Object one, two, three, four` (this is also supported by the fact that the prototype already uses the `Select-Object` cmdlet) , it just need to figure out what properties are used.

> even though the collect-all-input-up-front behavior is a departure

Some of the existing parameters of the `Select-Object` cmdlet already doing this like `-Last <int>` (e.g. `-Last 1000`) and (unexpected, see: #11221) `-Unique`

**Comment by mklement0** [comment]
@iRon7:

My concern about `-AllProperties` and `-EveryProperty` is that it sounds like `Select-Object *` and doesn't express the aspect of _unifying_ (making _uniform_) the set of properties across all objects. In general, it is not uncommon for parameter names to use verbs (`-Wait`, `-Skip`, `-Force`, ...), if that is your concern.

And to be clear: I'm fine with adding this to `Select-Object`, and the collect-all-input-first is just a matter of documenting it properly. 

If `-Last` really currently collects all input up front, it amounts to an inefficient implementation that should be changed in favor of a queue of the specified length, so that only the most recent N input objects are retained on an ongoing basis. 

Similarly,
`-Unique` only needs to retain the _unique_ input objects - though, _depending on the input_, that may be all of them.

To put it differently: cmdlets that for conceptual reasons must defer pipeline input until their `end` block is processed may need to _look_ at all input objects first, but they don't have to necessarily _collect_ them _all_, whereas the feature we're discussing here does.

**Comment by iRon7** [comment]
Caveat on prototype:
* An error occurs in case the property names are equal but differ in casing.

Based on [this StackOverflow answer](https://stackoverflow.com/a/41578530/1701026) and some testing, I think it is safe presume that a `HashSet` stays ordered for this application.

**Updated prototype version:**
```PowerShell
function UnifyProperties {
  $Names = [System.Collections.Generic.HashSet[string]]::new([StringComparer]::OrdinalIgnoreCase)
  $InputCollected = @($Input)
  $InputCollected.ForEach({ 
    foreach ($Name in $_.psobject.Properties.Name) { $Null = $Names.Add($Name) }
  })
  $inputCollected | Select-Object @($Names)
}

[pscustomobject] @{ one = 1; two = 2; three = 3 },
[pscustomobject] @{ ONE = 10; THREE = 30; FOUR = 4 } |
    UnifyProperties

one two three FOUR
--- --- ----- ----
  1   2     3
 10        30 4
```

**Comment by iRon7** [comment]
Maybe I should open a new issue for this idea as this view is towards the PowerShell engine rather than the (`Select-Object`) cmdlet.

Going back to the actual problem I am trying to (easily) resolve in issue, the question ***why*** the issue "[Not all properties displayed](https://stackoverflow.com/q/44428189/1701026)" exists should be asked. Which I think I might define as follows:

*The cause of the issue is adherent to a pipeline (streaming) design where you can't see with objects will follow. Therefore it can only be assumed that the following objects have the same properties as the first object...*

With lead to a counter question:
*So, if I don't use the pipeline, I won't have this issue?*

Indeed, if you don't stream you shouldn't have this issue. Yet in PowerShell, if I do load the whole stream in memory, I actually still have the issue.

Bringing this back to a PowerShell expectation:

If I use the pipeline (stream) and do e.g. this:
```PowerShell
[pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'} 
```
I might expect to lose some properties.
But if I choke the pipeline and load everything in memory by using the [grouping operator `( )`](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_operators#grouping-operator--):
```PowerShell
([pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'})
```
or assign the complete array to a variable:
```PowerShell
$Array = [pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'}
$Array
```
I might actually might expect that PowerShell has enough information available to unify all the properties automatically for me.

Technically, I assume this could be implemented as a "syntactic sugar" similar to the `UnifyProperties` prototype purpose but from performance point of view it probably better to build a list of unique PSObject names during the construction of the native PowerShell (PSObject) array.

**Comment by jhoneill** [comment]
The "why does PowerShell output the first object before waiting to see ALL the objects and their properties" question sort of answers itself. It Outputs the first and writes headers and then assumes anything that follows will be same, so you don't need to wait for all he data to come out of a long running process.   
 One of the interesting questions is would we want something like `format-table -property * -autosize ` which _does_ wait until it has all the objects to do this kind of unification ?  That might be a better place to make the change than in Select-Object (but the cmdlet isn't the only thing that uses the table formatter so that might be more of an engine thing) 
I can see an argument for doing it in `Select-object` -  that it is a common enough requirement that it should be in the core product, and I can see the arguments against: (a) it isn't hard to solve , and (b) it changes how select object works (storing objects rather and emitting them all at the end). (c) It's a formatting thing so any work should really be in the formatting engine. 
 
> But if I choke the pipeline and load everything in memory by using the [grouping operator `( )`]> 
> ```powershell
> ([pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'})
> ```
> 
> or assign the complete array to a variable:

Those things _still_ send a set of objects down the pipeline with no destination so they go through the formatter and plop out on the console.  Anything which sends multiple dissimilar objects in one command will do the same thing - output objects as they arrive, using the first to set the format for the others.

**Comment by iRon7** [comment]
@jhoneill,

> Those things still send a set of objects down the pipeline with no destination so they go through the formatter and plop out on the console

That is not just console cmdlets but any (destination) cmdlet in the pipeline, take e.g. `Export-Csv` or `ConvertTo-Csv`:
```PowerShell
[pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'} |ConvertTo-Csv
"name","address"
"joe","home"
,
```
Wishful thinking to resolve this:
```PowerShell
([pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'}) |ConvertTo-Csv
"name","address","phone"
"joe","home",
,,"1"
```

**Comment by jhoneill** [comment]
@iRon7  strictly it isn't any command. And I was responding to some sloppy wording: all output goes down the PowerShell pipeline.  You can write to host or verbose or error but that's not output. 

Many commands assume that all objects the receive will be of the same type Something like ConvertTO-Csv does it. Export-Excel (which I worked on) does it - and Doug Finke wrote something to do this job there too.  If you want to write a header with property types (excel, csv, but not JSON for example), then you have two options, either assume identical objects, write the header based on the first and carry on from there, which is typically faster and can move items on to the next step as they are ready. 
or wait until you have the last object, look at all the properties of all the objects to see if there are any new ones, and then output all the columns.  I don't think you're going to get everything to work this way by default. So you're back to putting a unify command in before the one which makes the assumptions

**Comment by iRon7** [comment]
Sorry to be persistent on this...

> ...or wait until you have the last object

if I use the [grouping operator `( )`](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_operators#grouping-operator--) or assign the object array (pipeline) to a variable, I do already have all the objects _up till last object_: `( ... )[-1]`.

Unfortunately I wasn't able to write a class (in PowerShell) that completely does what I have in mind, where I would like to align (unify) the items when they are being unrolled in the pipeline rather than at the construction. Nevertheless, they prototype I created appears to work quiet well the way I imaging it to work by default:

```PowerShell
class Unify : System.Collections.Generic.List[PSObject] {
    Unify([Object[]]$Objects) {
        ForEach($Object in $Objects) {
            $This.Add($Object)
            if ($Object -isnot [String]) {
                $Object.PSObject.Properties.Name.ForEach{ 
                    $This |Add-Member -MemberType ScriptProperty -Name $_ -Value {} -ErrorAction SilentlyContinue
                }
            }
        }
    }
}

[Unify]([pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'}) |% { $_ }

name address phone
---- ------- -----
joe  home
             1
```

It would be nice if PowerShell would be able to align (unify) the properties each object by default at moment it unrolls a (in-memory) array into a pipeline similar what the above class does (at construction).
Note that the individual items should still have only their own properties, e.g.:
```PowerShell 
([Unify]([pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'}))[1]

phone
-----
1

([Unify]([pscustomobject]@{name='joe';address='home'}, [pscustomobject]@{phone='1'}))[0..1]

name address phone
---- ------- -----
joe  home
             1
```

**Comment by jhoneill** [comment]
Grouping (per that link) doesn't mean what you think it means. 
`1 + 2 * 3 `    gives 7    
`(1 + 2) * 3` gives 9.   It _groups operations_ 
`1, 2 * 3`  Gives 1,2,1,2,1,2  (it repeats the array 3 times) 
`1, (2 * 3)`  Gives 1 , 6  (2* 3 is _grouped_ before making the array) 
`(1, 2) * 3`  is the default (make the array a _group_ - which was the default anyway - and then make 3 of those)

There is an _array operator_ so `(1)` is just the int 1 , but `@(1)`  is a single member array with int 1 as the sole member

If  PowerShell sees `1,2,3 | thing`   or `@(1,2,3) | thing`   or `(1,2,3) | thing`  it says "oh an array of objects, I'll enumerate its members and send them down the pipeline one by one... 
`$x | convertTo-csv`  does the the same as `($x) | convertTo-csv`   but `($x + $y) | convert-toCsv` is _grouped_ differently from `$x + ($y | convert-toCsv)` 

I see where you're going with that class, but you'll be quicker making an array, doing a pass through it to find the full list of properties. The do a second pass through it to add the missing ones. What you have there will get very slow with a big array.  Just pondering if what you actually want is a special type of array/collection which does that when items are enumerated or even as they are added.

**Comment by iRon7** [comment]
Please either mark this issue with something like <kbd>Resolution-Declined</kbd> or put it on a kind of <kbd>wishlist</kbd> "[*Issue we would like to prioritize, but we can't commit we will get to it yet*](https://github.com/dotnet/runtime/labels?q=wishlist)

**Comment by scotthardwick** [comment]
Has there been any movement on this?

**Comment by iRon7** [comment]
Please either mark this issue with something like <kbd>Resolution-Declined</kbd> or (knowing that it has several upvotes), put it on a kind of <kbd>wishlist</kbd> "[*Issue we would like to prioritize, but we can't commit we will get to it yet*](https://github.com/dotnet/runtime/labels?q=wishlist)

**Comment by iRon7** [comment]
Please either mark this issue with something like <kbd>Resolution-Declined</kbd> or put it on a kind of <kbd>wishlist</kbd> "[*Issue we would like to prioritize, but we can't commit we will get to it yet*](https://github.com/dotnet/runtime/labels?q=wishlist)

**Comment by minig0d** [comment]
Bump + naming suggestion: -UnionProperties 

I agree that the earlier suggestion of -Union alone was technically incorrect as it implies that items themselves are being unioned... but you were on the right track... what's actually being Unioned is the Property Names.... And this seems very consistent with existing ones.

I also completely agree that powershell should generate a warning because of the resultant unintended data loss...

**Comment by nascentt** [comment]
> This issue has not had any activity in 6 months, if there is no further activity in 7 days, the issue will be closed automatically.
. -->

Pity. I still heavily rely on this workaround code due to powershell's shortcomings.

**Comment by iRon7** [comment]
Is it possible to triage this issue and either reject (<kbd>Resolution-Declined</kbd>) the propose or something like:<kbd>Up-for-Grabs</kbd>/<kbd>KeepOpen</kbd>?

**Comment by nascentt** [comment]
Considering Microsoft made a breaking change this year with -UseBasicParsing now being a required field, I feel like adding -UnionProperties as an optional field is a no-brainer.

**Comment by iRon7** [comment]
Is it possible to triage this issue and either reject (<kbd>Resolution-Declined</kbd>) the propose or something like:<kbd>Up-for-Grabs</kbd>/<kbd>KeepOpen</kbd>?
